### PR TITLE
New: `max-lines` rule (fixes #6078)

### LINKED ISSUE #6078: Rule Proposal: max-lines
From [maximumNumberOfLines].

This rule will disallow that a file gets the number of lines more than the specific integer.

```js
{
    "max-lines": ["error", {
        "max": 300,
        "skipBlankLines": false,
        "skipComments": false
    }]
}
```

- `max` (default is `300`) - The maximum count of lines per file.
- `skipBlankLines` (default is `false`) - If this is `true`, this rule skips blank lines to count lines.
- `skipComments` (default is `false`) - If this is `true`, this rule skips comment lines to count lines.

[maximumNumberOfLines]: http://jscs.info/rule/maximumNumberOfLines

### PR DESCRIPTION
**Pull request by alberto**

### CODE REVIEW

File: docs/rules/max-lines.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+## Rule Details
+
+This rule enforces a maximum number of lines per file, in order to aid in maintainability and reduce complexity.
+
+
+## Options
+
+This rule has a number or object option:
+
+* `"max"` (default `100`) enforces a maximum number of lines in a file
```
Comment (TimvdLippe) [inline docs/rules/max-lines.md]: The default should be 300: https://github.com/eslint/eslint/pull/6321/files#diff-662109f5bdfe0d12503c019cf4ae3ed2R52

Comment (alberto) [inline docs/rules/max-lines.md]: Thanks @TimvdLippe good catch. I intended to leave it as it is in JSCS (100) when I noticed it was different but forgot to update the code.

@mysticatea why did you change it from 100 to 300? @eslint/eslint-team any opinions on what value this should be?

Comment (mikesherov) [inline docs/rules/max-lines.md]: Keep it at the same default as JSCS for maximum compatibility.

Comment (platinumazure) [inline docs/rules/max-lines.md]: That's an interesting question for larger discussion-- does JSCS compatibility mean match JSCS as closely as possible to make integration require less effort, or is ESLint allowed to have a different opinion on defaults and "JSCS compatibility" merely means there is a way to run your old JSCS settings in ESLint (even with different/non-default configuration)?

Comment (nzakas) [inline docs/rules/max-lines.md]: I think matching JSCS is the way to go.

File: lib/rules/max-lines.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
+/**
+ * @fileoverview enforce a maximum file length
+ * @author Alberto Rodríguez
+ */
+"use strict";
+
+var lodash = require("lodash");
+var astUtils = require("../ast-utils");
```
Comment (nzakas) [inline lib/rules/max-lines.js]: Can you add the requirements header comment here?

File: docs/rules/max-lines.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+```js
+/*eslint max-lines: ["error", 2, {"skipComments": true}]*/
+// a comment
+var a,
+    b, c;
+```
+
+## When Not To Use It
+
```
Comment (nzakas) [inline docs/rules/max-lines.md]: We've been adding a "Compatibility" section describing the JSCS rule that our rule relates to. Can you add that here? See http://eslint.org/docs/rules/no-useless-rename#compatibility

Comment (alberto) [inline docs/rules/max-lines.md]: Should I link to the original rule? I did a quick review and we are not very consistent with this.

A few examples:
http://eslint.org/docs/rules/object-property-newline#compatibility
http://eslint.org/docs/rules/linebreak-style#compatibility
http://eslint.org/docs/rules/indent#compatibility

Comment (pedrottimark) [inline docs/rules/max-lines.md]: 👍  to ordinary link to the JSCS rule page, similar to Related Rules, except that under Compatibility one of the following precedes the link: JSLint, JSHint, JSCS

Good catch, I will submit a pull request to make the inconsistent ones compatible 😀

File: docs/rules/max-lines.md (hunk 1 above)
Comment (mysticatea) [inline docs/rules/max-lines.md]: I'm sorry, I'm not sure why I changed it. No reason.

**Comment by alberto** [comment]
PR updated addressing comments above.

File: docs/rules/max-lines.md (hunk 1 above)
Comment (markelog) [inline docs/rules/max-lines.md]: There is no "default" for this rule in JSCS, what mentioned in the docs is just an example.

`100` feels weird to me, i'd say http://www.mind2b.com/component/content/article/24-software-module-size-and-file-size is interesting, my vote for default (if one should exist) is 500

Comment (alberto) [inline docs/rules/max-lines.md]: Oh, you are right, sorry for misleading here. We do have defaults for every rule, so yes, we have to choose one.

Comment (nzakas) [inline docs/rules/max-lines.md]: :+1: for 300

**Comment by alberto** [comment]
Default value updated to 300.

**Comment by markelog** [comment]
> Default value updated to 300.

Perhaps we can add some sort of explanation why we choose this number?

**Comment by nzakas** [comment]
Agree with @markelog, I think a little note and a reference to that link would go a long way.

**Comment by nzakas** [comment]
Lgtm
